Source/bmalloc/libpas/src/libpas/pas_stats.h

```diff
@@ -0,0 +1,311 @@
+#define PAS_STATS_MALLOC_INFO_LOW_GRANULARITY             (1ull << PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_SHIFT)
+#define PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_BUCKETS     (PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_MAX_SIZE / PAS_STATS_MALLOC_INFO_LOW_GRANULARITY)
+// FIXME: handle larger sizes in a sane way -- powers of two?
+#define PAS_STATS_MALLOC_INFO_OVERSIZE_BUCKETS            (1ull)
+
+#define PAS_STATS_MALLOC_INFO_BUCKET_COUNT_PER_STAT (\
+    PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY_BUCKETS + \
+    PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_BUCKETS + \
+    PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_BUCKETS + \
+    PAS_STATS_MALLOC_INFO_OVERSIZE_BUCKETS)
+
+typedef struct {
+    pas_stats_stat_base base;
+    uint64_t total_count;
+    uint64_t count_by_heap_type[pas_stats_heap_type_count];
+    uint64_t count_by_size[PAS_STATS_MALLOC_INFO_BUCKET_COUNT_PER_STAT];
+} pas_stats_malloc_info_data;
+
+#if PAS_ENABLE_STATS
+
+char* pas_stats_malloc_info_log(void* stat_v);
+void pas_stats_malloc_info_record(pas_stats_malloc_info_data* data, pas_stats_heap_type heap_type, size_t size, size_t count);
+
+#define PAS_STATS_FOR_EACH_COUNTER(OP) \
+    OP(malloc_info_bytes, pas_stats_malloc_info_data, pas_stats_malloc_info_log) \
+    OP(malloc_info_allocations, pas_stats_malloc_info_data, pas_stats_malloc_info_log)
+
+// FIXME: in principle it should be possible to automatically generate this via PAS_STATS_FOR_EACH_COUNTER, somehow
+#define PAS_RECORD_STAT_malloc_info_bytes(data, heap_type, size) \
+    pas_stats_malloc_info_record(data, heap_type, size, size)
+#define PAS_RECORD_STAT_malloc_info_allocations(data, heap_type, size) \
+    pas_stats_malloc_info_record(data, heap_type, size, 1)
+
+#else /* !PAS_ENABLE_STATS */
+#define PAS_RECORD_STAT(name, ...) do { PAS_UNUSED_V(__VA_ARGS__); } while (0)
+#endif /* PAS_ENABLE_STATS */
+
+#define PAS_RECORD_STAT_MALLOC(heap_type, size) \
+    PAS_RECORD_STAT(malloc_info_bytes, heap_type, size); \
+    PAS_RECORD_STAT(malloc_info_allocations, heap_type, size);
+
+/******************************************/
+/* Back to general stat-counter machinery */
+/******************************************/
+
+#if PAS_ENABLE_STATS
+
+void pas_stats_do_accounting_before_recording_stat(void);
+// Assumes g_pas_stats_data.log_lock is held
+void pas_stats_log_all_enabled_stats(void);
+
+// It's OK for the .enabled check here to be non-atomic: we know that all
+// .enabled bits will start initialized to 1, and will be set to 0 at most once.
+// If a thread fails to observe that write-to-0, the consequence is that it will
+// make a few unnecessary atomic writes to some stat-counters, but those stat
+// counters will never actually be used since pas_stats_log_all_enabled_stats
+// does use an atomic check for whether .enabled is set.
+#define PAS_RECORD_STAT(name, ...) do { \
+        if (g_pas_stats_data.per_stat_data. name .base.enabled) { \
+            pas_stats_do_accounting_before_recording_stat(); \
```

Comment:
Instead of incrementing `g_pas_stats_data.log_counter` once per stat type, let's add a bool in `pas_stats_data` which is set to true if any of the underlying stats is true.  Then, have `PAS_RECORD_STAT_MALLOC` check this bool and increment `g_pas_stats_data.log_counter` once if needed.  Do this before invoking the respective `PAS_RECORD_STAT`s.  Each `PAS_RECORD_STAT` need not call `pas_stats_do_accounting_before_recording_stat()` themselves.  This will minimize the perf cost when no stats collection is enabled.

Reply:
I expect most other stat-counters to be 'solitary', i.e. incremented on their own -- at least among the ones I have in mind to add. If that changes I'll want to create a proper 'stat group' layer over the base data level, but until then I don't want to complicate the code too much in service of 'paired' stats like this.

For this case, I think we can achieve the same perf savings by adding a PAS_RECORD_STAT_WITHOUT_ACCOUNTING macro which would skip the accounting step, and then making PAS_RECORD_STAT_MALLOC use that for the second call.